File: schemas/digital-asset-management/drizzle/asset_tags.ts
// asset_tags: Join table associating tags with assets.
// See README.md for full design rationale.

import { pgTable, uuid, timestamp, index, unique } from "drizzle-orm/pg-core";
import { users } from "../../auth-rbac/drizzle/users";
import { assets } from "./assets";
import { tags } from "./tags";

export const assetTags = pgTable(
  "asset_tags",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    assetId: uuid("asset_id").notNull().references(() => assets.id, { onDelete: "cascade" }),
    tagId: uuid("tag_id").notNull().references(() => tags.id, { onDelete: "cascade" }),
    assignedBy: uuid("assigned_by").notNull().references(() => users.id, { onDelete: "restrict" }),
    assignedAt: timestamp("assigned_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_asset_tags_asset_id_tag_id").on(table.assetId, table.tagId),
    index("idx_asset_tags_tag_id").on(table.tagId),
  ]
);

File: schemas/digital-asset-management/spacetimedb/asset_tags.rs
// asset_tags: Join table linking assets to tags.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

// Composite unique(asset_id, tag_id) — enforce in reducer logic

#[spacetimedb::table(name = asset_tags, public)]
pub struct AssetTag {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub asset_id: String, // FK → assets.id (cascade delete)

    #[index(btree)]
    pub tag_id: String, // FK → tags.id (cascade delete)

    pub assigned_by: String, // FK → users.id (restrict delete)

    pub assigned_at: Timestamp,
}
